esp/services/ws_dali/ws_daliservice.cpp

```diff
@@ -96,3 +239,50 @@ void CWSDaliSoapBindingEx::exportSDSData(CHttpRequest* request, CHttpResponse* r
     io.clear();
     removeFileTraceIfFail(outFileNameWithPath);
 }
+
+int CWSDaliSoapBindingEx::onStartUpload(IEspContext &ctx, CHttpRequest* request, CHttpResponse* response, const char *serv, const char *method)
+{
+    StringArray fileNames, files;
+    VStringBuffer source("WsDali::%s()", method);
+    Owned<IMultiException> me = MakeMultiException(source);
+    try
+    {
+#ifdef _USE_OPENLDAP
+        request->queryContext()->ensureSuperUser(ECLWATCH_SUPER_USER_ACCESS_DENIED, "Access denied, administrators only.");
+#endif
+        if (wsdService->isDaliDetached())
+            throw makeStringException(ECLWATCH_CANNOT_CONNECT_DALI, "Dali detached.");
+
+        if (strieq(method, "Import"))
+        {
+            StringBuffer path, add;
+            request->getParameter("Path", path);
+            request->getParameter("Add", add);
+            if (isEmptyString(path))
+                throw makeStringException(ECLWATCH_INVALID_INPUT, "Data path not specified.");
+
+            request->readContentToFiles(nullptr, daliFolder, fileNames);
+            unsigned count = fileNames.ordinality();
+            if (count == 0)
+                throw MakeStringException(ECLWATCH_INVALID_INPUT, "Failed to read upload content.");
+            //For now, we only support importing 1 file per Import request for a better response time.
+            //Some file could be very big. It may take a log time to import.
+            if (count > 1)
+                throw MakeStringException(ECLWATCH_INVALID_INPUT, "Only one file is allowed.");
+
+            VStringBuffer fileName("%s%s", daliFolder, fileNames.item(0));
+            wsdService->importSDSData(nullptr, fileName, path, strToBool(add));
```

Comment:
would it be better perhaps, if the service dealt with XML only (not file), and the file was loaded to XML here, instead of conditionally inside importSDSData?

Reply:
Would it be more efficient to create a tree directly using createPTreeFromXMLFile() comparing with the 2 steps: read the XML string from a file and, then, call createPTreeFromXMLString()?